src/PasswordPolicy.php

```diff
@@ -0,0 +1,289 @@
+<?php
+
+namespace Drupal\password_enhancements;
+
+use Drupal\password_enhancements\Exception\PolicyInvalidArgumentException;
+use Drupal\user\RoleInterface;
+
+/**
+ * Defines the Password Policy.
+ */
+final class PasswordPolicy implements PasswordPolicyInterface {
+
+  /**
+   * Expiry of the password in seconds, 0 means that passwords doesn't expire.
+   *
+   * @var int
+   */
+  private $expireSeconds;
+
+  /**
+   * Show warning message before the password would expire in a given seconds.
+   *
+   * @var int
+   */
+  private $expireWarnSeconds;
+
+  /**
+   * Expiry warning message.
+   *
+   * @var string|null
+   */
+  private $expiryWarningMessage;
+
+  /**
+   * The number of constraints that are required.
+   *
+   * @var int
+   */
+  private $minimumRequiredConstraints;
+
+  /**
+   * Role ID.
+   *
+   * @var string
+   */
+  private $role;
+
+  /**
+   * The array of password constraints for this password policy.
+   *
+   * @var array
+   */
+  private $constraints = [];
+
+  /**
+   * Holds the collection of password constraints that are used by this policy.
+   *
+   * @var \Drupal\password_enhancements\PasswordConstraintPluginCollection
+   */
+  private $constraintsCollection;
+
+  /**
+   * PasswordPolicy constructor.
+   *
+   * @param \Drupal\user\RoleInterface $role
+   *   The role this Password Policy belongs to.
+   * @param int $expire_seconds
+   *   The expiry in seconds.
+   * @param int $expire_warn_seconds
+   *   The expiry warning in seconds.
+   * @param string|null $expiry_warning_message
+   *   The expiry warning message.
+   * @param int $minimum_required_constraints
+   *   The minimum number of required constraints.
+   * @param array $constraints
+   *   The constraints of the policy.
+   *
+   * @phpcs:disable DrupalPractice.Objects.UnusedPrivateMethod.UnusedMethod
+   *   PHPCS fails to recognize that static::createFromRole() calls this.
+   */
+  private function __construct(RoleInterface $role, int $expire_seconds, int $expire_warn_seconds, ?string $expiry_warning_message, int $minimum_required_constraints, array $constraints) {
```

Comment:
Maybe the real question here whether it worths to differentiate `NULL` from an empty string or not. I mean, in both cases, the message should not be displayed to a user because it does not contain any information.

Reply:
That's not 100% correct. The real question is whether we should make the warning message required when the policy's `expire_seconds` AND `expire_warn_seconds` are both greater than 0 or not. If the answer is yes to this question, then the original question vanishes.